### PR TITLE
Implement primitives in Source 3 Concurrent

### LINKED ISSUE #1161: Improve function is_stream
using new arity function, see https://github.com/source-academy/js-slang/issues/1160

### LINKED ISSUE #1171: Source §3 Concurrent: primitive functions missing
char_at: not declared
display_list: not declared
draw_data: declared but always returns undefined
[arity: new function, see https://github.com/source-academy/js-slang/pull/1170; will have the same problem]

although these functions are listed here: https://docs.sourceacademy.org/source_3_concurrent/

What should happen:

char_at: should work as usual
display_list, draw_data: would be nice, but it's ok if it would be declared and return the same as display, but with a message "(limited implementation)"

### PR DESCRIPTION
**Pull request by teekaytai**
This pull request resolves #1171 by implementing display_list, char_at, arity and draw_data in S3 Concurrent.
(It also resolves #1161 by swapping out array_length for arity in every definition of is_stream)

display_list: Implemented in a manner similar to display

char_at: Implemented by calling char_at from stdlib/misc

arity: Implemented in the machine. Note that the way arity behaves differs between Concurrent and regular versions of Source for variadic functions. In vanilla Source, it returns the number of parameters other than the rest parameter. In the concurrent version, since all variadic functions are indicated to have `VARARGS_NUM_ARGS = -1` arguments, there is no way to see how many compulsory arguments the function has. I have set it to return 0 for any variadic function. This may have unintended consequences if one expects a function with arity 0 to be callable with 0 arguments (like in is_stream).
```
arity(display)  // 1 in Source 3
arity(display)  // 0 in Source 3 Concurrent
```
draw_data: No longer returns undefined after adding it back to the array of `EXTERNAL_PRIMITIVES`. Also ensured the return result matches that of draw_data in normal Source (when draw_data receives exactly one argument, it returns the argument as is after drawing it, otherwise it returns an array containing its arguments)
```
draw_data(pair(1, 2))  // [1, 2]
draw_data(pair(1, 2), pair(3 4))  // [[1, 2], [3, 4]]
```

### CODE REVIEW

**Review by thomastanck — COMMENTED** [review]
Since these are new primitives, we would need to add the new primitive ids to https://github.com/source-academy/js-slang/wiki/SVML-Instruction-Set#primitive-functions if this gets merged.

I don't mind adding these new primitives other than needing to align the instruction set with the C implementation as well: https://github.com/source-academy/sinter

File: src/vm/__tests__/svml-machine.ts [hunk 1]
```diff
@@ -646,6 +646,27 @@ describe('primitive opcodes', () => {
       )
     })
 
+    // Also note the result in visualiseListResult in the .snap file
```
[the remark this answers is not in the document]
Comment (teekaytai) [reply src/vm/__tests__/svml-machine.ts:649]: I'll leave that to you then. Thanks

[the remark this answers is not in the document]
Comment (martin-henz) [reply src/vm/__tests__/svml-machine.ts:649]: So @thomastanck you are making the changes you have in mind by pushing to this PR?

[the remark this answers is not in the document]
Comment (thomastanck) [reply src/vm/__tests__/svml-machine.ts:649]: Sorry, I have been away for the past couple weeks. I'll try to make the changes on Wednesday!

[the remark this answers is not in the document]
Comment (martin-henz) [reply src/vm/__tests__/svml-machine.ts:650]: @thomastanck any update on this?

[the remark this answers is not in the document]
Comment (thomastanck) [reply src/vm/__tests__/svml-machine.ts:650]: Sorry I didn't ping you, it's ready for merge if you're happy with the changes!

**Review by thomastanck — APPROVED** [review]

**Comment by martin-henz** [comment]
Oh, I just realized: `draw_data` returns an array in Source §2. That is weird because Source §2 doesn't have arrays. Not sure what is better. Maybe a list?

The current solution will be confusing because students do not know arrays yet and cannot parse [1, 2, 3], the result of `draw_data(1, 2, 3);`.

**Comment by martin-henz** [comment]
Here are the options for `draw_data`:
* One argument: return the argument; 0, 2 or more: return a list with the arguments
* Any number of arguments: return the list of arguments
* One argument: return the argument; 0, 2 or more: return undefined
* Any number of arguments: return the first argument
I think the last option is the best: It's the most homogeneous, the easiest to implement and it is useful: You can still use applications of `draw_data` in expressions, and you get a result that makes sense. This solution is also closest to the behaviour of `display` and `display_list`, which also returns the first argument, regardless whether the second argument is present or not.

**Comment by thomastanck** [comment]
I have implemented your suggestion, but there is a quirk where draw_data(x) returns x as opposed to list(x). Should draw_data(x) also return list(x)?

**Comment by martin-henz** [comment]
> I have implemented your suggestion, but there is a quirk where draw_data(x) returns x as opposed to list(x). Should draw_data(x) also return list(x)?

Messages crossed.

**Comment by thomastanck** [comment]
I like the comparison with display. draw_data's return value is most useful in expressions, and in that usecase we should prioritise making the behaviour of draw_data(x) most intuitive (i.e. return x). Having draw_data(x, y) return x as well would match draw_data(x)'s behaviour better than returning a list.

**Comment by martin-henz** [comment]
> I like the comparison with display. draw_data's return value is most useful in expressions, and in that usecase we should prioritise making the behaviour of draw_data(x) most intuitive (i.e. return x). Having draw_data(x, y) return x as well would match draw_data(x)'s behaviour better than returning a list.

In that case, we should insist that draw_data has at least one argument. A call `draw_data()` should raise an error.

**Comment by thomastanck** [comment]
Sounds good to me!

**Comment by martin-henz** [comment]
> Sounds good to me!

OK, do you want to implement? Probably a trivial change.

**Comment by thomastanck** [comment]
Nearly done implementing! Will push in about 15 minutes.

**Pull request merged**